### PR TITLE
Disable needless specialization in Artifacts code

### PR DESCRIPTION
**Pull request by staticfloat**
After spending some quality time with `--trace-compile=stderr` and `@snoopi`, I have found a fairly minimal set of changes that speeds up loading of [`TestGtkMegaJLL.jl`](https://github.com/giordano/TestGtkMegaJLL.jl) from `2.12s` to `0.69s` on my machine.  It's all essentially nonspecialization hacks.

### CODE REVIEW

**Comment by staticfloat** [comment]
@KristofferC I tried simplifying the artifact code to eliminate the download path, but it didn't actually help anything, so I elected for the smaller changes (with no change in behavior, which that would constitute for 1.3.1)

**Comment by staticfloat** [comment]
The cumulative impact of all the optimizations we've done so far is that `using GTK3_jll` has gone from `12.5s` on 1.3.0 to `6.1s` on this branch.  Not bad!

**Comment by KristofferC** [comment]
What's the status here? Should it be merged for 1.3.1?

**Comment by staticfloat** [comment]
This should eventually be merged into 1.3.X (and 1.4 of course); it's just performance improvements so it's not time-critical, but it's a nice-to-have.  There's nothing critical missing in this PR; although the test isn't useful yet, because I don't know how to make the precompile tester actually fail, so our precompile script can still bitrot.

**Comment by StefanKarpinski** [comment]
Isn’t improving the package load time on 1.3 fairly pressing?

**Comment by KristofferC** [comment]
Yeah, but I know Jameson didn't like us putting more Pkg precompile into the sysimage. But I think we should merge this and put it in 1.3.1 and keep thinking about how it can be improved.

**Comment by KristofferC** [comment]
The way to check the precompilation script is to check that the output after each command is as expected.

File: src/Pkg.jl [hunk 1]
```diff
@@ -537,6 +537,17 @@ const precompile_script = """
     ] add Te\t\t$CTRL_C
     ] st
     $CTRL_C
+    # Create simple artifact, bind it, then use it:
+    foo_hash = Pkg.Artifacts.create_artifact(dir -> touch(joinpath(dir, "foo")))
+    Pkg.Artifacts.bind_artifact!("./Artifacts.toml", "foo", foo_hash)
+    # Also create multiple platform-specific ones because that's a codepath we need precompiled
+    Pkg.Artifacts.bind_artifact!("./Artifacts.toml", "foo_plat", foo_hash; platform=platform_key_abi())
```
Comment (KristofferC) [inline src/Pkg.jl]: Did this every work?

```
julia> using Pkg.Artifacts

julia> platform_key_abi
ERROR: UndefVarError: platform_key_abi not defined
```

Comment (giordano) [reply src/Pkg.jl]: It's in Pkg.BinaryPlatforms if I remember correctly

Comment (KristofferC) [reply src/Pkg.jl]: Doesn't seem like it..?

Comment (staticfloat) [reply src/Pkg.jl]: Perhaps it got lost in a rebase somewhere; I distinctly remember adding a `using Pkg.BinaryPlatforms` as well as testing this and having it work.

**Pull request merged**